src/fastpm.c

```diff
@@ -730,11 +736,40 @@ prepare_ncdm(FastPMSolver * fastpm, RunData * prr, MPI_Comm comm)
         shift0 = 0;
     }
     double shift[3] = {shift0, shift0, shift0};
-    
+
     // fill the ncdm store to make a grid with nc_ncdm grid points in each dim
     ptrdiff_t Nc_ncdm[3] = {nc_ncdm, nc_ncdm, nc_ncdm}; 
-    fastpm_store_fill(fastpm_solver_get_species(fastpm, FASTPM_SPECIES_NCDM), fastpm->pm, shift, Nc_ncdm);
+    fastpm_store_fill(ncdm_sites, fastpm->pm, shift, Nc_ncdm);
+
+    // stagger the ncdm grid wrt the cdm grid. FIXME: Does this conflict with the shift stuff above?
+    int i, d;
+    for(i = 0; i < ncdm_sites->np; i ++) {
+        for(d = 0; d < 3; d ++) {
+            ncdm_sites->x[i][d] += fastpm->config->boxsize / nc_cdm * 0.5;
+            if (ncdm_sites->q)
+                ncdm_sites->q[i][d] += fastpm->config->boxsize / nc_cdm * 0.5;
+        }
+    }
+
+    // SPLIT
+    fastpm_split_ncdm(nid, ncdm_sites, ncdm, comm);
+
+    // replace ncdm species with the split ncdm
+    fastpm_solver_add_species(fastpm,
+                              FASTPM_SPECIES_NCDM,
+                              ncdm);
+
+    /* after splitting, the ncdm particles have moved so we must
+       apply periodic boundary and move particles to the correct rank */
+    PM * pm = fastpm->pm;   // FIXME: I'm not sure this is what we want it to be. I think this is overwritten a lot throughout the solver.
+    int NTask;
+    MPI_Comm_size(fastpm->comm, &NTask);
     
+    fastpm_store_wrap(ncdm, BoxSize);
+    fastpm_store_decompose(ncdm,
+                           (fastpm_store_target_func) FastPMTargetPM, pm,
+                           fastpm->comm);
+
```

Comment:
@rainwoodman We've changed things so that splitting neutrinos also involves displacing the particles. This means we have to wrap the particles after splitting. I have also decomposed the particles using fastpm->pm. Is this necessary at the IC stage, and also will this work as wanted? I'm not sure if fastpm->pm is what we want it to be?

Reply:
Yes. I agree with you.